[filesystem] Improve performance over high-latency network links

riviti opened the pull request:
## Description

Rework FileCache so that it tries to read data in bigger blocks,
but as a multiple of the IFile::GetChunkSize() value. IFile documents
that GetChunkSize() should return the smallest suitable/efficient
read size, so this still follows that API, but it allows IFile
implementations more freedom to optimize.

In the case of (posix) SMB, this allows the samba library to issue
much bigger read commands which results in more data getting transferred
for every network roundtrip. The protocol supports reads of up to
8MiB, but the samba library will send multiple commands if the
read size is bigger than this. Hence, almost any read size is supported,
which means that GetChunkSize() can return 1 now without fear
of getting a bunch of 1 byte read calls.

With these changes, it is possible to stream high bitrate 1080p files
accessed over SMB across a WAN+VPN link, instead of just over a LAN.

## Motivation and context

I am not aware of an open issue, but I have not looked. I found an itch and scratched it.

## How has this been tested?

Lightly tested thus far, on x86_64 Linux, only with an SMB media source. In this case a remote SMB server accessed over a VPN. Some more testing would be good, especially for NFS shares. I'll see what/if I can manage more thorough testing unless this is simply immediately NAK'd.

## What is the effect on users?

Enhanced possibilities of consuming content also in poor(er) network conditions.

## Screenshots (if appropriate):
N/A

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed (exept for TestWebServer.CanHeadFile which seems broken even without my changes)

riviti committed:
[filesystem] Improve performance over high-latency network links

Rework FileCache so that it tries to read data in bigger blocks,
but as a multiple of the IFile::GetChunkSize() value. IFile documents
that GetChunkSize() should return the smallest suitable/efficient
read size, so this still follows that API, but it allows IFile
implementations more freedom to optimize.

In the case of (posix) SMB, this allows the samba library to issue
much bigger read commands which results in more data getting transferred
for every network roundtrip. The protocol supports reads of up to
8MiB, but the samba library will send multiple commands if the
read size is bigger than this. Hence, almost any read size is supported,
which means that GetChunkSize() can return 1 now without fear
of getting a bunch of 1 byte read calls.

With these changes, it is possible to stream high bitrate 1080p files
accessed over SMB across a WAN+VPN link, instead of just over a LAN.

lrusak commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -234,8 +234,36 @@ void CFileCache::Process()
+    // consumes over a given time period, so during 500ms we must
+    // read 10MiB. This is obviously a simplified view of the network,
+    // but for now it's better than nothing.
+    requestSize = 10 * 1024 * 1024;
```
Please add this as a constexpr int in an anonymous namespace at the start of the file.

```
namespace
{

constexpr int DEFAULT_REQUEST_SIZE = 10 * 1024 * 1024;

}
```

lrusak commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -234,8 +234,36 @@ void CFileCache::Process()
+
   // create our read buffer
-  std::unique_ptr<char[]> buffer(new char[m_chunkSize]);
+  std::unique_ptr<char[]> buffer(new char[requestSize]);
```
This looks like a good candidate for `std::vector<uint8_t>` (possibly out of the scope of this PR and not a required change at all).

lrusak commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -312,25 +340,30 @@ void CFileCache::Process()
-    if (maxSourceRead > 0)
-      iRead = m_source.Read(buffer.get(), maxSourceRead);
+    CLog::Log(LOGDEBUG, "CFileCache::{} Reading {} bytes from source. maxWrite: {}",
+              __FUNCTION__, maxSourceRead, maxWrite);
```
```suggestion
              __func__, maxSourceRead, maxWrite);
```

wsnipex replied:
TODO: this is way too spammy, so either it should be component logging or removed before merge

riviti replied:
I see a potential need for it in the future, how do I go about making it component logging? I tried grepping a little bit but didn't quite figure out what I need to do wrt the UI list of components.

wsnipex replied:
May bad, component logging is only used for extra logging from external libraries.
Not quite sure how to deal with the spam here, maybe this can only be logged when we're about to run dry and skip it when all is good? Other thoughts?

riviti replied:
Something might be possible.

We have the m_writeRate that we might use to figure out if we're having trouble keeping up, perhaps.

Or only logging when the buffer is almost empty like you said.

riviti replied:
Hopefully it's less spammy now, at least for operation where there is some headroom.

With my large-file-over-VPN it still takes about a minute until it stops printing. Not a huge issue there as that is about one line per second, but it could be worse for non-SMB sources.

Tell me what you think!

lrusak commented on xbmc/platform/posix/filesystem/SMBFile.h:
```diff
@@ -79,7 +79,7 @@ class CSMBFile : public IFile
   bool Delete(const CURL& url) override;
   bool Rename(const CURL& url, const CURL& urlnew) override;
-  int GetChunkSize() override { return 64*1024; }
+  int GetChunkSize() override { return 1; }
```
This seems odd. I read the description in the OP but I still don't really understand this.

wsnipex replied:
they way I understand it: samba doesn't care about chunk size, so by returning 1 here, file cache is free to read as much data as possible from samba in one request.

riviti replied:
For reference, here's the `GetChunkSize()` doc comment (from IFile.h):
```
  /* Returns the minimum size that can be read from input stream.  *
   * For example cdrom access where access could be sector based.  *
   * This will cause file system to buffer read requests, to       *
   * to meet the requirement of CFile.                             *
   * It can also be used to indicate a file system is non buffered *
   * but accepts any read size, have it return the value 1         */
  virtual int  GetChunkSize() {return 0;}

```
I interpret that like it should return 1 for SMBFile, because it has no special requirements regarding read size. Before this commit however that would have caused FileCache to try to read one byte at a time. Well not really, because the now removed `CFile::DetermineChunkSize()` call in would have used the advancedsettings memorysize instead... 

64kiB was the maximum read request in SMB < 2.1 I believe, so possibly that is the origin of the old value, or maybe it was just a round number.

riviti replied:
@wsnipex Exactly!

In practice a 64kiB chunk size is also ok I believe, in my environment the reads have been a few MiBs each with this change so 64kiB just rounds that down slightly, maybe 1-5%

riviti replied:
I may have changed my mind a little bit. One thing that the old block size did was to ensure that FileCache doesn't do too *small* reads. Now it can do very small ones if the storage is fast. So maybe that value should be left alone after all.

lrusak reviewed (commented):
Thanks for the PR!

I don't have much knowledge about the code in question (and I'm just doing the review on my phone) so I can't comment too much on the functional changes. I've added a couple comments about style and possible improvements though.

riviti commented:
@lrusak For the condition formatting suggestions, and the `__FUNCTION__` -> `__func__` thing, I didn't see anything specific in the style guide and the current formatting matches the file.

I can definitely change it, but given that information, should I?

riviti commented:
I'm a little concerned that this change might slow down things like thumbnailing due to issuing an excessively large read when only a small amount is needed.

I'm not sure how to fix that, as the FileCache class doesn't seem to really know how much the calling code wants to read.

One thing might be to try to start small and successively read larger and larger blocks of data until things stabilize (that way the first few hundred bytes would be ready much sooner), but I'm concerned that such an algorithm might be difficult to tune so that it reliably picks a large enough read size.

arnova commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -136,9 +136,9 @@ bool CFileCache::Open(const CURL& url)
-      m_source.GetChunkSize(),
-      CServiceBroker::GetSettingsComponent()->GetAdvancedSettings()->m_cacheChunkSize);
+  m_chunkSize = m_source.GetChunkSize();
+  if (!m_chunkSize)
```
@lrusak :Well actually it should be "if (m_chunkSize == 0)" (without the !)

arnova commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -312,25 +349,30 @@ void CFileCache::Process()
+    // over network file systems will hopefully increase throughput.
+    // Special case: m_chunkSize == 1 (non-buffered file) fits
+    // right into this calculation.
+    int64_t maxSourceRead = maxWrite - maxWrite % m_chunkSize;
```
Please use () for clarity

riviti replied:
I haven't added them, but the code looks different now anyway, please tell me if you still think it's needed.

arnova commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -234,8 +252,27 @@ void CFileCache::Process()
+  {
+    // We have "unlimited" buffer capabilities, but we still must bound the
+    // request size to something.
+    requestSize = DEFAULT_REQUEST_SIZE;
```
I don't think this will work for streaming filesystems like ftp/http/webdav. Afaik the read size there is limited by the internal ringbuffer in CCurlFile. I think the assumption you're doing here is solely based on what smb can do...

riviti replied:
Yes, only SMB has been tested, as noted.

I can test other things and for sure there's likely to be at least some problem somewhere. Nevertheless I don't see a way to improve performance for the filesystem case without increasing the read size somehow.

Regarding `IFile`s which can't deliver this much data at a time, would it be acceptable to change those so that they return less data in those cases? Roughly matching the behaviour of the posix `read()` syscall?

wsnipex replied:
I've quickly tested this with NFS and HTTP on LAN and didn't notice any immediate issues.

arnova replied:
Ah yes that's probably because we handle big reads in CCurlFile::ReadData(). But still: I think it's a bad idea for at least streamed filesystems, especially since there's not much gain for those doing big reads. 

Wouldn't it be better to simply add something like MaxReadSize() to IFile so filesystems can tell how much they can return at once?

riviti replied:
That might not be a bad thing to have, but there's quite a few implementations to add it to and I'm not sure a default value can be picked in a reasonable way since it depends on the bitrate. IMO, ideally it should be based on time and the bitrate of the media tuned so we get for example 1-100 read calls per second. I often deal with raw `read()` syscalls so to me letting `Read()` return less than requested feels like second nature to me.

In either case, CCurlFile seems to return short reads sometimes and that currently works (at least before these changes). That doesn't say anything about other types of course but the rest of the code seems to cope with those short reads at least.

I've tried to test the Youtube and TED Talk plugins but those seem broken in master so I didn't get to any `IFile` things that way.

riviti replied:
@wsnipex Thanks for your testing!

I'm wondering if the NFS code should have `chunkSize = 1` too with similar arguments as for SMBFile. I might dig into that.

arnova replied:
> That might not be a bad thing to have, but there's quite a few implementations to add it to and I'm not sure a default value can be picked in a reasonable way since it depends on the bitrate. IMO, ideally it should be based on time and the bitrate of the media tuned so we get for example 1-100 read calls per second. I often deal with raw `read()` syscalls so to me letting `Read()` return less than requested feels like second nature.
> 
> In either case, CCurlFile seems to return short reads sometimes and that currently works (at least before these changes). That doesn't say anything about other types of course but the rest of the code seems to cope with those short reads at least.
> 
> I've tried to test the Youtube and TED Talk plugins but those seem broken in master so I didn't get to any `IFile` things that way.

Please keep in mind that, especially with stuff on the internet, when a large read (partially) fails, we need to possibly read a larger block all over again. This is also a reason why you don't want too large reads especially with streaming filesystems. And of course larger reads also require more memory, but that's probably not an issue on most hardware.

I'd make the implementation in a way that it behaves the same as it did before except for the filesystems of which we know they allow larger reads and where it's beneficial, like smb/nfs.

wsnipex replied:
libnfs bounds chunk size to whatever the server supports.

riviti replied:
Normally I would expect the posix `read()` to deal with partial failures by first returning what succeeded and then returning an error or retrying on the next call. Dealing with block boundaries complicate it somewhat and you still have a point about only fixing what's broken. I'll give extending `IFile` a try so we can see what it looks like!

riviti replied:
Ok, there's an implementation now. It's a little bit hairy in the `CFileCache` class, but overall not too bad I think.

arnova commented:
@riviti : I assume you didn't test this with any streaming filesystem like http/ftp/webdav ? I strongly doubt the new logic will work properly for those (and possibly others as well).

riviti committed:
[filesystem] Use std::make_unique<T> in FileCache

riviti committed:
[filesystem] Add IFile::GetMaxReadSize() and use it to tune caching

riviti commented on xbmc/filesystem/File.cpp:
```diff
@@ -945,9 +950,16 @@ int CFile::IoControl(EIoControl request, void* param)
+{
+    if (m_pFile)
+      return m_pFile->GetMaxReadSize();
+    return SpecialReadSizeRequest::INVALID;
```
Pff. Maybe I should set my editor up properly...

riviti commented:
I'll fix the jenkins formatting suggestions, but I'll keep them local for now to reduce the noice in the PR.

Risca commented on xbmc/filesystem/IFile.cpp:
```diff
@@ -78,6 +78,27 @@ bool IFile::ReadString(char *szLine, int iLineLength)
+      // Backwards compatibility:
+      // CFileCache used to interpret a chunk size of 0 or 1 as
+      // 1) Get chunk size from advanced settings
+      // 2) Read one chunk at a time (the only way it knew how!)
```
Don't you mean
```c
/// 0) Foo
/// 1) Bar
```
?

riviti replied:
No, it was the same two parts to the handling for both chunk size 0 and 1. But I see how that might be confusing.

riviti replied:
Clearer?
```
    case 0:
    case 1:
      // Backwards compatibility:
      // CFileCache used to handle a chunk size of 0 or 1 by
      // a) Getting the chunk size from advanced settings
      // b) Reading one chunk at a time (the only way it knew how!)
      // To not change this behavior for IFile implementations,
      // Assume the ONE_CHUNK behavior in those cases and let the IFiles
      // that want something different override this method.
      return SpecialReadSizeRequest::ONE_CHUNK;
```

Risca replied:
Much better! :smiley:

thexai commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -235,7 +292,7 @@ void CFileCache::Process()
 
   // create our read buffer
-  std::unique_ptr<char[]> buffer(new char[m_chunkSize]);
+  auto buffer = std::make_unique<std::vector<char>>(m_readSize);
```
It seems that the natural way to handle a buffer with std::vector should be:

In FileCache.h
`std::vector<char> m_buffer;`

Here just:
`m_buffer.resize(m_readSize);`

riviti replied:
Since the thread is persistent for the life of the `CFileCache` instance I'm not sure it'd make a whole lot of difference, either way we're creating exactly one vector (but in your example potentially allocating the internal buffer twice, at creation and then at resize).

Style-wise I personally prefer to keep it local since it's a smaller scope.

thexai replied:
> I personally prefer to keep it local since it's a smaller scope.

Then maybe:
`std::vector<char> buffer(m_chunkSize);`

riviti replied:
Yes, that makes sense!

thexai commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -312,25 +369,36 @@ void CFileCache::Process()
 
-    const int64_t maxWrite = m_pCache->GetMaxWriteSize(m_chunkSize);
-    int64_t maxSourceRead = m_chunkSize;
+    const int64_t maxWrite = m_pCache->GetMaxWriteSize(buffer->size());
```
`m_buffer.size()`

thexai commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -312,25 +369,36 @@ void CFileCache::Process()
-      iRead = m_source.Read(buffer.get(), maxSourceRead);
+    CLog::Log(LOGDEBUG, "CFileCache::{} Reading {} bytes from source. maxWrite: {}",
+              __FUNCTION__, maxSourceRead, maxWrite);
+    ssize_t iRead = m_source.Read(buffer->data(), maxSourceRead);
```
`m_buffer.data()`

arnova commented on xbmc/filesystem/IFile.h:
```diff
@@ -39,6 +39,14 @@ class CURL;
 namespace XFILE
 {
 
+enum SpecialReadSizeRequest
```
Can we rename this to something different? Perhaps something like ReadSizeRequestCode ?

riviti replied:
Sure!

riviti committed:
[filesystem] FileCache: Make cache level diagnostic conditional

For small chunk sizes the diagnostic is potentially logged a lot, so
only log in certain cases where we might suspect there's a problem.

riviti committed:
[filesystem] FileCache: Read slowly in the beginning and after seeks

Attempt to make a series of smaller reads starting when there is no
data in the buffer to hopefully get to a playable amount of data quickly.

riviti commented:
> I'm a little concerned that this change might slow down things like thumbnailing due to issuing an excessively large read when only a small amount is needed.
> 
> I'm not sure how to fix that, as the FileCache class doesn't seem to really know how much the calling code wants to read.
> 
> One thing might be to try to start small and successively read larger and larger blocks of data until things stabilize (that way the first few hundred bytes would be ready much sooner), but I'm concerned that such an algorithm might be difficult to tune so that it reliably picks a large enough read size.

I've tried that solution now.

It seems to work ok on a PC, but I'm not sure how it'd work on something slow with a hardware decode engine like a Raspberry pi. Getting to the "next cache refill" takes longer and longer now which might mean some sort of stall in the decoding pipeline perhaps?

I'm going to try to get a raspberry pi build going (and hopefully find some hardware...) but I'm starting from zero there.

wsnipex commented on xbmc/filesystem/FileCache.cpp:
```diff
@@ -312,25 +370,53 @@ void CFileCache::Process()
+    // Update m_slowStartReadSize for the next round. Once it has reached
+    // buffer.size() it has served it's purpose.
+    if (static_cast<size_t>(m_slowStartReadSize) < buffer.size())
+	m_slowStartReadSize *= 2;
```
this multiplier would be a candidate for an advanced setting to allow user tuning.
possibly also read limit.

mrfixit2001 commented:
For what it's worth - I tested a few variations of this PR backported to Matrix on an ARM RK3399 (GBM + DRM) over WLAN. In every test it actually hindered the ability for smooth video playback. I tested a few large 1080p files as well as a few 4k files, all with the same result.

riviti commented:
This has been rotting for too long, I should try to make a push to get it into a better state :/

> For what it's worth - I tested a few variations of this PR backported to Matrix on an ARM RK3399 (GBM + DRM) over WLAN. In every test it actually hindered the ability for smooth video playback. I tested a few large 1080p files as well as a few 4k files, all with the same result.

@mrfixit2001 It sounds odd that it would become worse. Do you have any logs? My best guess would be that the reads are still small in your situation, for whatever reason and maybe some part of the auto tuning doesn't do the right thing. Could you try to use wireshark or such to figure out

1) the SMB version
2) the sizes of the SMB read requests

Or get a pcap file for me to look at, for example using tcpdump on your target hardware? Consider the privacy implications before doing so though. SMB can do encryption but it's much harder to inspect that way.

Or edit this code to always log (it will potentially be verbose enough to cause trouble on it's own if the read sizes are small): https://github.com/xbmc/xbmc/pull/20775/commits/3d6461bcb2a68efd62c931edbbf24d90618cba8f

Otherwise maybe you could push your branch somewhere so I can try it. I don't have access to the same hardware however.
